### PR TITLE
Test if auth host is reachable

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
This PR removes a conservative host name validation using a plain regex. Instead we test and report if the host (optionally with port) is reachable.

Screen Shot 2022-07-15 at 12 24 06

## Related Issue(s)

Improves the situation around #5305

## How to test

Try configuring and using a Git Provider which is listening on a port other than the default 443, thus requires to specify a port. 

## Release Notes

```release-note
Test if host of a Git Integration is reachable.
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by roboquat** [comment]
@AlexTugarev: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/server/src/auth/auth-provider-service.ts [hunk 1]
```diff
@@ -172,7 +173,12 @@ export class AuthProviderService {
...
 
     protected callbackUrl = (host: string) => {
-        const pathname = `/auth/${host}/callback`;
+        const safeHost = host.replace(":", "_");
+        const pathname = `/auth/${safeHost}/callback`;
         return this.config.hostUrl.with({ pathname }).toString();
     };
+
+    async isHostReachable(host: string) {
+        return await isReachable(host, { timeout: 2000 });
```
Comment (geropl) [inline components/server/src/auth/auth-provider-service.ts:182]: :100:

File: components/server/src/auth/auth-provider-service.ts [hunk 2]
```diff
@@ -172,7 +173,12 @@ export class AuthProviderService {
     }
 
     protected callbackUrl = (host: string) => {
-        const pathname = `/auth/${host}/callback`;
+        const safeHost = host.replace(":", "_");
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/server/src/auth/auth-provider-service.ts:176]: that `callbackUrl` goes into settings of the auth provider (runtime part) which means, it needs to be stable. 

thanks for the hint, there is a handler side to it as well, we need to see if the same adjustment needs to be made there as well.

**Comment by AlexTugarev** [comment]
BTW I tagged it `security` relevant, as I was wondering about any chances to override existing saas providers like gihub.com with something like github.com:433 and somehow interfere with authorizations and basically enable scamming.

I tested and verified that it's not possible to use github.com:433, gitlab.com:433 ✅ 

cc. @geropl

**Comment by jldec** [comment]
Nice - this is a helpful improvement. Thanks @AlexTugarev.

**Review by MrSimonEmms — APPROVED** [review]
I've done a few tests on it and it's working as expected - good work.

I created a GitLab instance on http://20.0.96.8:81/ and https://gitlab.<mydomain>.com and was able to verify the connection ok. **NB** the IP address wasn't able to login as we require HTTPS, but the domain one worked fine.

**Comment by AlexTugarev** [comment]
/unhold

I see no risk in just trying this out.

**Comment by lucasvaltl** [comment]
👋 This is awesome and was a very quick turnaround! One question: Did we add something to the docs to indicate that this is now possible - for example in [this](https://www.gitpod.io/docs/gitlab-integration) section?

**Pull request merged**